File: remote/java/src/main/java/examples/TypeTest.java
package examples;

import java.math.BigInteger;
import java.net.InetAddress;
import java.net.URI;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;
import org.apache.log4j.Logger;
import org.pxscene.rt.RTEnvironment;
import org.pxscene.rt.RTException;
import org.pxscene.rt.RTFunction;
import org.pxscene.rt.RTObject;
import org.pxscene.rt.RTValue;
import org.pxscene.rt.RTValueType;
import org.pxscene.rt.remote.RTRemoteConnectionManager;
import org.pxscene.rt.remote.RTRemoteMulticastResolver;

/**
 * test all types, set type with new value and check get value equals old value.
 */
public class TypeTest {

  private static final Logger logger = Logger.getLogger(TypeTest.class);

  /**
   * the total examples
   */
  private int totalExamplesCount = 0;

  /**
   * the suceed examples
   */
  private int succeedExamplesCount = 0;

  public static void main(String[] args) throws Exception {

    RTEnvironment.init();
    RTRemoteMulticastResolver resolver = new RTRemoteMulticastResolver(
        InetAddress.getByName("224.10.10.12"),
        10004);
    URI uri = resolver.locateObject("host_object");
    RTObject obj = RTRemoteConnectionManager.getObjectProxy(uri);
    TypeTest rtRemoteTestClient = new TypeTest();
    while (true) {
      try {
        rtRemoteTestClient.setTotalExamplesCount(0);
        rtRemoteTestClient.setSucceedExamplesCount(0);
        rtRemoteTestClient.doFunctionTest(obj, "onTick");
        rtRemoteTestClient.doObjectTest(obj, "objvar");
        rtRemoteTestClient.testAllTypes(obj);
        logger.debug("test completed, next test will at 10s ...");
        Thread.sleep(1000 * 10);
      } catch (Exception err) {
        err.printStackTrace();
      }
    }
  }

  /**
   * test all basic type
   * float,bool,int8,uint8,int32,uint32,int64,uint64,double,string,void ptr
   */
  public void testAllTypes(RTObject rtObject)
      throws RTException, ExecutionException, InterruptedException {
    logger.debug("======== start test rtValue type =======");

    // float last 7 digit
    doBasicTest(rtObject, RTValueType.FLOAT, 12.0123123f, "ffloat");
    doBasicTest(rtObject, RTValueType.FLOAT, -123.8818f, "ffloat");
    doBasicTest(rtObject, RTValueType.FLOAT, 199123123.91f, "ffloat");

    //test bool
    doBasicTest(rtObject, RTValueType.BOOLEAN, true, "bbool");
    doBasicTest(rtObject, RTValueType.BOOLEAN, false, "bbool");

    // int8 range [-128,127]
    doBasicTest(rtObject, RTValueType.INT8, (byte) -128, "int8");
    doBasicTest(rtObject, RTValueType.INT8, (byte) 0, "int8");
    doBasicTest(rtObject, RTValueType.INT8, (byte) 127, "int8");

    //test uint8, the data range is[0,255]
    doBasicTest(rtObject, RTValueType.UINT8, (short) 0, "uint8");
    doBasicTest(rtObject, RTValueType.UINT8, (short) 255, "uint8");

    //test int32, range is 	[–2147483648 , 2147483647]
    doBasicTest(rtObject, RTValueType.INT32, -2147483648, "int32");
    doBasicTest(rtObject, RTValueType.INT32, 0, "int32");
    doBasicTest(rtObject, RTValueType.INT32, 123, "int32");
    doBasicTest(rtObject, RTValueType.INT32, 2147483647, "int32");

    // test uint32, range is [0 - 4,294,967,295]
    doBasicTest(rtObject, RTValueType.UINT32, 0L, "uint32");
    doBasicTest(rtObject, RTValueType.UINT32, 4294967295L, "uint32");
    doBasicTest(rtObject, RTValueType.UINT32, 123123L, "uint32");

    // test int64, range is [–9223372036854775808  9223372036854775807]
    doBasicTest(rtObject, RTValueType.INT64, -9223372036854775808L, "int64");
    doBasicTest(rtObject, RTValueType.INT64, 9223372036854775807L, "int64");
    doBasicTest(rtObject, RTValueType.INT64, 0L, "int64");
    doBasicTest(rtObject, RTValueType.INT64, 123123L, "int64");

    // test uint64, range is [0 - 18446744073709551615]
    doBasicTest(rtObject, RTValueType.UINT64, new BigInteger("18446744073709551615"), "uint64");
    doBasicTest(rtObject, RTValueType.UINT64, BigInteger.valueOf(0), "uint64");
    doBasicTest(rtObject, RTValueType.UINT64, BigInteger.valueOf(123123123), "uint64");
    doBasicTest(rtObject, RTValueType.UINT64, BigInteger.valueOf(123), "uint64");

    // test double
    doBasicTest(rtObject, RTValueType.DOUBLE, 1231.12312312312, "ddouble");
    doBasicTest(rtObject, RTValueType.DOUBLE, -1231.12312312312, "ddouble");
    doBasicTest(rtObject, RTValueType.DOUBLE, -0.12, "ddouble");

    // test string
    doBasicTest(rtObject, RTValueType.STRING,
        "implemented in both /Library/Java/JavaVirtualMachines/jdk1.8.0_40.jdk/Conten", "string");
    doBasicTest(rtObject, RTValueType.STRING,
        "{\"jsonKey\":\"values\"}", "string");
    doBasicTest(rtObject, RTValueType.STRING, "1", "string");

    // void ptr is a uint32 or uint64
    doBasicTest(rtObject, RTValueType.VOIDPTR, 723123231L, "vptr");
    doBasicTest(rtObject, RTValueType.VOIDPTR, 789892349L, "vptr");

    logger.debug(String
        .format("========= %d of %d example succeed, %d failed.", succeedExamplesCount,
            totalExamplesCount, totalExamplesCount - succeedExamplesCount));
  }

  /**
   * do basic test examples
   *
   * @param rtObject the rt remote Object
   * @param type the rtObject type
   * @param value the rtValue
   * @param propertiesName the properties name
   */
  private void doBasicTest(RTObject rtObject, RTValueType type, Object value,
      String propertiesName) throws RTException, ExecutionException, InterruptedException {
    rtObject.set(propertiesName, new RTValue(value, type)).get();
    Object newVal = rtObject.get(propertiesName).get().getValue();
    totalExamplesCount += 1;

    boolean result;
    if (type == RTValueType.FLOAT) {
      result = checkEqualsFloat(value, newVal);
    } else if (type == RTValueType.DOUBLE) {
      result = checkEqualsDouble(value, newVal);
    } else {
      result = value.equals(newVal);
    }
    if (result) {
      succeedExamplesCount += 1;
    }
    logger.debug(getTestResult(type.toString(), value, newVal, result));
  }


  /**
   * test function type
   */
  private void doFunctionTest(RTObject rtObject, String propertiesName)
      throws RTException, ExecutionException, InterruptedException {

    RTValue oldRtValue = new RTValue(new RTFunction(rtValueList -> {
      logger.debug("doFunctionTest test...");
      logger.debug(rtValueList);
    }));

    rtObject.set(propertiesName, oldRtValue).get();
    Future<RTValue> valueFuture = rtObject.get(propertiesName);
    RTValue rtValue = valueFuture.get();
    ((RTFunction) rtValue.getValue()).getListener().invoke(null);
    totalExamplesCount += 1;
    boolean result = checkEqualsFunction(oldRtValue, rtValue);
    if (result) {
      succeedExamplesCount += 1;
    }

    logger.debug(
        getTestResult(RTValueType.FUNCTION.toString(), oldRtValue.getValue(),
            rtValue, result));
  }

  /**
   * test object type
   */
  private void doObjectTest(RTObject rtObject, String propertiesName)
      throws RTException, ExecutionException, InterruptedException {
    TestObject oldObj = new TestObject("type test");
    RTValue oldRtValue = new RTValue(oldObj, RTValueType.OBJECT);
    rtObject.set(propertiesName, oldRtValue).get();
    RTValue newObj = rtObject.get(propertiesName).get();

    TestObject testObject = (TestObject) newObj.getValue();
    testObject.doTest();
    totalExamplesCount += 1;

    boolean result = testObject.getId().equals(oldObj.getId());
    if (result) {
      succeedExamplesCount += 1;
    }
    logger.debug(
        getTestResult(RTValueType.OBJECT.toString(), oldObj.getId(),
            testObject.getId(), result));
  }

  /**
   * check function is equals
   */
  private boolean checkEqualsFunction(RTValue v1, RTValue v2) {
    return ((RTFunction) v1.getValue()).getFunctionName()
        .equals(((RTFunction) v2.getValue()).getFunctionName());
  }

  /**
   * floating point values can be off by a little bit, so they may not report as exactly equal.
   * so i need use eps to check equal
   */
  private boolean checkEqualsFloat(Object v1, Object v2) {
    float eps = 0.001f;
    if (Math.abs((float) v1 - (float) v2) < eps) {
      return true;
    }
    return false;
  }

  /**
   * double  values can be off by a little bit, so they may not report as exactly equal.
   */
  private boolean checkEqualsDouble(Object v1, Object v2) {
    double eps = 0.0001;
    if (Math.abs((double) v1 - (double) v2) < eps) {
      return true;
    }
    return false;
  }

  /**
   * get test result to debug
   *
   * @param type the test type
   * @param old the old value (set value)
   * @param newValue the new value (get from remote)
   * @param result the equals result
   */
  private String getTestResult(String type, Object old, Object newValue, boolean result) {
    return type + " test => set val = " + old + ", rpc result = " + newValue
        + ", passed = ["
        + result + "]";
  }

  public int getTotalExamplesCount() {
    return this.totalExamplesCount;
  }

  public int getSucceedExamplesCount() {
    return this.succeedExamplesCount;
  }

  public void setTotalExamplesCount(int totalExamplesCount) {
    this.totalExamplesCount = totalExamplesCount;
  }

  public void setSucceedExamplesCount(int succeedExamplesCount) {
    this.succeedExamplesCount = succeedExamplesCount;
  }
}

File: remote/js/examples/typeTest.js
/**
 * Copyright (C) 2018 TopCoder Inc., All Rights Reserved.
 */

/**
 * this file contains all type test examples
 *
 * @author      TCSCODER
 * @version     1.0
 */

const RTRemoteMulticastResolver = require('../lib/RTRemoteMulticastResolver');
const RTRemoteConnectionManager = require('../lib/RTRemoteConnectionManager');
const RTValueHelper = require('../lib/RTValueHelper');
const RTValueType = require('../lib/RTValueType');
const RTConst = require('../lib/RTConst');
const logger = require('../lib/common/logger');
const helper = require('../lib/common/helper');
const BigNumber = require('bignumber.js');

/**
 * the RTRemoteMulticastResolver instance
 * @type {RTRemoteMulticastResolver}
 */
const resolve = new RTRemoteMulticastResolver('224.10.10.12', 10004);

/**
 * the total number of examples
 * @type {number}
 */
let total = 0;

/**
 * the successful number of examples
 * @type {number}
 */
let succeed = 0;

resolve.start()
  .then(() => resolve.locateObject('host_object')) // find remote object
  .then(uri => RTRemoteConnectionManager.getObjectProxy(uri)) // crate remote object
  .then((rtObject) => {
    const doTest = () => {
      total = 0;
      succeed = 0;

      Promise.resolve() // test all type in sequence mode

      // function test
        .then(() => doFunctionTest(rtObject, 'onTick'))

        // object test
        .then(() => doObjectTest(rtObject, 'objvar'))

        // in c++/java, float only had 7 valid digits
        .then(() => doBasicTest(rtObject, RTValueType.FLOAT, 1.23456789, 'ffloat'))
        .then(() => doBasicTest(rtObject, RTValueType.FLOAT, -1.234567, 'ffloat'))
        .then(() => doBasicTest(rtObject, RTValueType.FLOAT, 12.34567, 'ffloat'))
        .then(() => doBasicTest(rtObject, RTValueType.FLOAT, 123.4567, 'ffloat'))
        .then(() => doBasicTest(rtObject, RTValueType.FLOAT, 1234.567, 'ffloat'))
        .then(() => doBasicTest(rtObject, RTValueType.FLOAT, 123456.71, 'ffloat'))

        // test bool
        .then(() => doBasicTest(rtObject, RTValueType.BOOLEAN, true, 'bbool'))
        .then(() => doBasicTest(rtObject, RTValueType.BOOLEAN, false, 'bbool'))

        // int8 range [-128,127]
        .then(() => doBasicTest(rtObject, RTValueType.INT8, -128, 'int8'))
        .then(() => doBasicTest(rtObject, RTValueType.INT8, 0, 'int8'))
        .then(() => doBasicTest(rtObject, RTValueType.INT8, 127, 'int8'))

        // test uint8, the data range is[0,255]
        .then(() => doBasicTest(rtObject, RTValueType.UINT8, 0, 'uint8'))
        .then(() => doBasicTest(rtObject, RTValueType.UINT8, 255, 'uint8'))

        // test int32, range is  [–2147483648 , 2147483647]
        .then(() => doBasicTest(rtObject, RTValueType.INT32, -2147483648, 'int32'))
        .then(() => doBasicTest(rtObject, RTValueType.INT32, 0, 'int32'))
        .then(() => doBasicTest(rtObject, RTValueType.INT32, 123, 'int32'))
        .then(() => doBasicTest(rtObject, RTValueType.INT32, 2147483647, 'int32'))

        // test uint32, range is [0 - 4,294,967,295]
        .then(() => doBasicTest(rtObject, RTValueType.UINT32, 0, 'uint32'))
        .then(() => doBasicTest(rtObject, RTValueType.UINT32, 4294967295, 'uint32'))
        .then(() => doBasicTest(rtObject, RTValueType.UINT32, 123123, 'uint32'))

        // test int64, range is [–9223372036854775808  9223372036854775807]
        .then(() => doBasicTest(rtObject, RTValueType.INT64, new BigNumber('-9223372036854775808'), 'int64'))
        .then(() => doBasicTest(rtObject, RTValueType.INT64, new BigNumber('9223372036854775807'), 'int64'))
        .then(() => doBasicTest(rtObject, RTValueType.INT64, new BigNumber(0), 'int64'))
        .then(() => doBasicTest(rtObject, RTValueType.INT64, new BigNumber(123123), 'int64'))

        // test uint64, range is [0 - 18446744073709551615]
        .then(() => doBasicTest(rtObject, RTValueType.UINT64, new BigNumber('18446744073709551615'), 'uint64'))
        .then(() => doBasicTest(rtObject, RTValueType.UINT64, new BigNumber(0), 'uint64'))
        .then(() => doBasicTest(rtObject, RTValueType.UINT64, new BigNumber(123123123), 'uint64'))
        .then(() => doBasicTest(rtObject, RTValueType.UINT64, new BigNumber('123'), 'uint64'))

        // test double
        .then(() => doBasicTest(rtObject, RTValueType.DOUBLE, 1231.12312312312, 'ddouble'))
        .then(() => doBasicTest(rtObject, RTValueType.DOUBLE, -1231.12312312312, 'ddouble'))
        .then(() => doBasicTest(rtObject, RTValueType.DOUBLE, -0.12, 'ddouble'))

        // test string
        .then(() => doBasicTest(
          rtObject, RTValueType.STRING,
          'implemented in both /Library/Java/JavaVirtualMachines/jdk1.8.0_40.jdk/Conten', 'string',
        ))
        .then(() => doBasicTest(
          rtObject, RTValueType.STRING,
          '{"jsonKey":"values"}', 'string',
        ))
        .then(() => doBasicTest(rtObject, RTValueType.STRING, '1', 'string'))

        // void ptr is a uint32 or uint64
        .then(() => doBasicTest(rtObject, RTValueType.VOIDPTR, new BigNumber('723123231'), 'vptr'))
        .then(() => doBasicTest(rtObject, RTValueType.VOIDPTR, new BigNumber('789892349'), 'vptr'))
        .then(() => {
          logger.debug(`========= ${succeed} of ${total} example succeed, ${total - succeed} failed.`);
          logger.debug('test completed, next test will at 10s ...');
          setTimeout(doTest, 10 * 1000);
        })
        .catch(err => logger.error(err));
    };
    doTest();
  })
  .catch((err) => {
    logger.error(err);
  });

/**
 * floating point values can be off by a little bit, so they may not report as exactly equal.
 * so i need use eps to check equal
 */
function checkEqualsFloat(v1, v2) {
  const eps = 0.001;
  return Math.abs(v1 - v2) < eps;
}

/**
 * double values can be off by a little bit, so they may not report as exactly equal.
 */
function checkEqualsDouble(v1, v2) {
  const eps = 0.0001;
  return Math.abs(v1 - v2) < eps;
}

/**
 * do basic test example
 * @param rtObject the remote object
 * @param type the rtValue type
 * @param value the value
 * @param propertyName the property name
 * @return {Promise<void>} the promise with void
 */
function doBasicTest(rtObject, type, value, propertyName) {
  return rtObject.set(propertyName, RTValueHelper.create(value, type)).then(() => rtObject.get(propertyName).then((rtValue) => {
    let result = false;
    if (type === RTValueType.FLOAT) {
      result = checkEqualsFloat(rtValue.value, value);
    } else if (type === RTValueType.DOUBLE) {
      result = checkEqualsDouble(rtValue.value, value);
    } else if (type === RTValueType.UINT64 || type === RTValueType.INT64 || type === RTValueType.VOIDPTR) {
      result = value.toString() === rtValue.value.toString(); //
    } else {
      result = rtValue.value === value;
    }
    printResult(type, value, rtValue.value, result);
  })).catch((err) => {
    logger.error(err);
  });
}

/**
 * print result and add total/success example number
 * @param type the rtValue type
 * @param old the old value
 * @param newValue the rpc value
 * @param result the test result
 */
function printResult(type, old, newValue, result) {
  total += 1;
  succeed += result ? 1 : 0;
  logger[result ? 'debug' : 'error'](`${helper.getTypeStringByType(type)} test => set val = ${old}, rpc result = ${
    newValue}, passed = [${result}]`);
}

/**
 * do function test
 * @param rtObject the remote object
 * @param propertyName the property name
 * @return {Promise<void>} the promise with void
 */
function doFunctionTest(rtObject, propertyName) {
  const oldRtValue = RTValueHelper.create((rtValueList) => {
    logger.debug('doFunctionTest test...');
    logger.debug(rtValueList);
  }, RTValueType.FUNCTION);

  return rtObject.set(propertyName, oldRtValue).then(() => rtObject.get(propertyName).then((rtValue) => {
    rtValue.value(null);
    const result = oldRtValue[RTConst.FUNCTION_KEY] === rtValue[RTConst.FUNCTION_KEY];
    printResult(RTValueType.FUNCTION, oldRtValue[RTConst.FUNCTION_KEY], rtValue[RTConst.FUNCTION_KEY], result);
  }));
}

/**
 * do object test
 * @param rtObject the remote object
 * @param propertyName the property name
 * @return {Promise<void>} the promise with void
 */
function doObjectTest(rtObject, propertyName) {
  const testObj = {
    hello: () => {
      logger.debug('hello from test obj');
    },
  };

  const rtOldObj = RTValueHelper.create(testObj, RTValueType.OBJECT);
  return rtObject.set(propertyName, rtOldObj).then(() => rtObject.get(propertyName).then((rtValue) => {
    const oldObjId = rtOldObj[RTConst.VALUE][RTConst.OBJECT_ID_KEY];
    const newObjId = rtValue[RTConst.VALUE][RTConst.OBJECT_ID_KEY];
    rtValue.value.hello();
    printResult(RTValueType.OBJECT, oldObjId, newObjId, oldObjId === newObjId);
  }));
}
